File: scripts/fetch-voxel51-dataset-zoo.mjs
import {access, mkdir, writeFile} from "node:fs/promises";
import path from "node:path";
import {fileURLToPath} from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const repoRoot = path.resolve(__dirname, "..");
const outputDir = path.join(repoRoot, "data", "voxel51-dataset-zoo");
const imageDir = path.join(outputDir, "images");

// Primary source for the built-in Voxel51 Dataset Zoo documentation.
const docUrl = process.env.VOXEL51_DATASET_ZOO_URL || "https://raw.githubusercontent.com/voxel51/voxel51-docs/main/docs/data/dataset_zoo/datasets.md";
const detailFields = ["Dataset name", "Dataset source", "Dataset size", "Tags", "Supported splits"];
const detailFieldHandlers = {
    "Dataset name": (section, value) => { section.datasetName = value; },
    "Dataset source": (section, value) => { section.source = value; },
    "Dataset size": (section, value) => { section.size = value; },
    "Tags": (section, value) => { section.tags = value.split(",").map(item => item.trim()).filter(Boolean); },
    "Supported splits": (section, value) => { section.supportedSplits = value.split(",").map(item => item.trim()).filter(Boolean); },
};
const docsResponse = await fetch(docUrl);
if (!docsResponse.ok) {
    throw new Error(`Failed to fetch Voxel51 docs: ${docsResponse.status} ${docsResponse.statusText}`);
}
const markdown = await docsResponse.text();

await mkdir(imageDir, {recursive: true});

const datasetSections = parseDatasetSections(markdown);
const datasets = [];

for (const section of datasetSections) {
    const imageUrls = Array.from(new Set(section.imageUrls));
    const savedImages = [];

    for (const imageUrl of imageUrls) {
        const imageResponse = await fetch(imageUrl);
        if (!imageResponse.ok) {
            console.warn(`[voxel51-zoo] failed to download image (${imageResponse.status} ${imageResponse.statusText}): ${imageUrl}`);
            continue;
        }
        const buffer = Buffer.from(await imageResponse.arrayBuffer());
        const filename = path.basename(new URL(imageUrl).pathname);
        // Keep the image filename deterministic while avoiding empty or malformed names.
        const safeFilename = filename.replace(/[^a-zA-Z0-9._-]/g, "_").replace(/^\.+/, "_");
        const baseFilename = safeFilename || "image";
        const imagePath = await getUniqueImagePath(imageDir, baseFilename);
        await writeFile(imagePath, buffer);
        savedImages.push({
            sourceUrl: imageUrl,
            localPath: path.relative(repoRoot, imagePath).split(path.sep).join("/"),
        });
    }

    datasets.push({
        id: slugify(section.name),
        name: section.name,
        summary: section.summary,
        datasetName: section.datasetName,
        source: section.source,
        size: section.size,
        tags: section.tags,
        supportedSplits: section.supportedSplits,
        notes: section.notes,
        imageUrls: imageUrls,
        images: savedImages,
        sourceUrl: section.sourceUrl,
    });
}

await writeFile(
    path.join(outputDir, "datasets.json"),
    JSON.stringify({
        generatedAt: new Date().toISOString(),
        sourceUrl: docUrl,
        datasets,
    }, null, 2) + "\n"
);

await writeFile(
    path.join(outputDir, "README.md"),
    buildReadme(datasets, docUrl)
);

async function getUniqueImagePath(imageDir, baseFilename) {
    const extension = path.extname(baseFilename);
    const stem = path.basename(baseFilename, extension);
    let candidate = path.join(imageDir, baseFilename);
    let counter = 1;

    while (true) {
        try {
            await access(candidate);
            const nextName = `${stem}-${counter}${extension}`;
            candidate = path.join(imageDir, nextName);
            counter += 1;
        } catch {
            return candidate;
        }
    }
}

function parseDatasetSections(markdown) {
    const lines = markdown.split(/\r?\n/);
    const sections = [];
    const detailPattern = new RegExp(`^-\\s+(${detailFields.join("|")}):\\s*(.+)$`);
    let current = null;

    for (const line of lines) {
        const headingMatch = line.match(/^##\s+(.+?)\s*(?:\[¶\]\(.*\))?\s*$/);
        if (headingMatch) {
            if (current) {
                sections.push(finalizeSection(current));
            }
            current = {
                name: headingMatch[1].trim(),
                summary: "",
                datasetName: "",
                source: "",
                size: "",
                tags: [],
                supportedSplits: [],
                notes: [],
                imageUrls: [],
                sourceUrl: "",
                paragraphs: [],
            };
            continue;
        }

        if (!current) {
            continue;
        }

        if (isLikelyParagraph(line)) {
            const paragraph = line.trim();
            if (paragraph) {
                current.paragraphs.push(paragraph);
            }
            continue;
        }

        const detailMatch = line.match(detailPattern);
        if (detailMatch) {
            const [, label, value] = detailMatch;
            const normalizedValue = normalizeMarkup(value);
            const handler = detailFieldHandlers[label];
            if (handler) {
                handler(current, normalizedValue);
            }
            continue;
        }

        const noteMatch = line.match(/^-\s+(.+)$/);
        if (noteMatch && !line.startsWith("- Dataset") && !current.notes.includes(noteMatch[1])) {
            current.notes.push(noteMatch[1]);
            continue;
        }

        const imageMatch = line.match(/!\[[^\]]*\]\(([^)]+)\)/);
        if (imageMatch) {
            current.imageUrls.push(resolveImageUrl(imageMatch[1], docUrl));
            continue;
        }
    }

    if (current) {
        sections.push(finalizeSection(current));
    }

    return sections.filter(section => section.name && section.name !== "Built-In Zoo Datasets");
}

function finalizeSection(section) {
    const summary = section.paragraphs.join(" ").trim() || section.summary || "";
    const normalizedNotes = section.notes
        .map(item => normalizeMarkup(item))
        .filter(Boolean);
    return {
        ...section,
        summary,
        notes: normalizedNotes,
        sourceUrl: `https://docs.voxel51.com/user_guide/dataset_zoo/datasets.html#${buildDocAnchor(section.name)}`,
    };
}

function buildDocAnchor(value) {
    return slugify(value);
}

function isLikelyParagraph(line) {
    const trimmed = line.trim();
    if (!trimmed) return false;
    if (trimmed.startsWith("**")) return false;
    if (trimmed.startsWith("![")) return false;
    if (trimmed.startsWith("-")) return false;
    return !trimmed.startsWith("Note") && !trimmed.startsWith("##");
}

function normalizeMarkup(value) {
    return value
        .replace(/`/g, "")
        .replace(/\[(.*?)\]\((.*?)\)/g, "$1")
        .replace(/\s+/g, " ")
        .trim();
}

function resolveImageUrl(imagePath, baseUrl) {
    return new URL(imagePath, baseUrl).toString();
}

function slugify(value) {
    return value
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, "-")
        .replace(/(^-|-$)/g, "");
}

function buildReadme(datasets, docUrl) {
    const headers = [
        "# Voxel51 Dataset Zoo catalog",
        "",
        "This directory stores a machine-readable catalog of the built-in Voxel51 FiftyOne Dataset Zoo datasets, plus preview images downloaded from the official docs source.",
        "",
        "## Source",
        "",
        `- Official docs: ${docUrl}`,
        "",
        "## Datasets",
        "",
        ...datasets.map(dataset => `- **${dataset.name}** — ${dataset.summary || "No summary available."} (${dataset.images.length} image${dataset.images.length === 1 ? "" : "s"})`),
    ];

    return headers.join("\n") + "\n";
}

File: scripts/fetch_voxel51_dataset_zoo.py
#!/usr/bin/env python3
import json
import posixpath
import re
from pathlib import Path
from typing import Any

import requests

REPO_URL = "https://raw.githubusercontent.com/voxel51/voxel51-docs/main/docs/data/dataset_zoo/datasets.md"
RAW_BASE = "https://raw.githubusercontent.com/voxel51/voxel51-docs/main/"
ROOT = Path(__file__).resolve().parents[1]
OUTPUT_DIR = ROOT / "data" / "voxel51"
IMAGES_DIR = OUTPUT_DIR / "images"
CATALOG_PATH = OUTPUT_DIR / "dataset-zoo-catalog.json"
SUMMARY_PATH = OUTPUT_DIR / "dataset-zoo-summary.md"


def download_text(url: str) -> str:
    response = requests.get(url, timeout=60)
    response.raise_for_status()
    return response.text


def download_binary(url: str, destination: Path) -> Path:
    if destination.exists():
        return destination
    destination.parent.mkdir(parents=True, exist_ok=True)
    response = requests.get(url, timeout=60)
    response.raise_for_status()
    destination.write_bytes(response.content)
    return destination


def slugify(value: str) -> str:
    value = re.sub(r"[^a-z0-9]+", "-", value.lower()).strip("-")
    return value or "dataset"


def parse_dataset_sections(markdown: str) -> list[tuple[str, str]]:
    pattern = re.compile(r"^##\s+(.*)$", re.MULTILINE)
    matches = list(pattern.finditer(markdown))
    sections: list[tuple[str, str]] = []
    for index, match in enumerate(matches):
        heading = match.group(1).strip()
        start = match.end()
        end = matches[index + 1].start() if index + 1 < len(matches) else len(markdown)
        content = markdown[start:end]
        sections.append((heading, content))
    return sections


def extract_details(section_content: str) -> dict[str, str]:
    details: dict[str, str] = {}
    for line in section_content.splitlines():
        stripped = line.strip()
        if stripped.startswith("- Dataset name:"):
            details["dataset_name"] = re.sub(r"[`*]", "", stripped.split(":", 1)[1]).strip()
        elif stripped.startswith("- Dataset source:"):
            source_match = re.search(r"\[(.*?)\]\((.*?)\)", stripped)
            details["source"] = source_match.group(2) if source_match else stripped.split(":", 1)[1].strip()
        elif stripped.startswith("- Dataset size:"):
            details["size"] = stripped.split(":", 1)[1].strip()
        elif stripped.startswith("- Tags:"):
            tags_raw = stripped.split(":", 1)[1].strip()
            details["tags"] = tags_raw.strip("`")
        elif stripped.startswith("- Supported splits:"):
            details["splits"] = stripped.split(":", 1)[1].strip()
    return details


def extract_description(section_content: str) -> str:
    pre_details = section_content.split("**Details**", 1)[0]
    cleaned = re.sub(r"!\[[^\]]*\]\(([^)]+)\)", "", pre_details)
    cleaned = re.sub(r"\[([^\]]+)\]\(([^)]+)\)", r"\1", cleaned)
    cleaned = re.sub(r"[`*_#]", "", cleaned)
    paragraphs = [p.strip() for p in re.split(r"\n\s*\n", cleaned) if p.strip()]
    if not paragraphs:
        return ""
    description = re.sub(r"\s+", " ", paragraphs[0]).strip()
    return description[:800]


def extract_images(section_content: str) -> list[str]:
    image_refs = re.findall(r"!\[[^\]]*\]\(([^)]+)\)", section_content)
    resolved: list[str] = []
    for ref in image_refs:
        ref = ref.strip()
        if not ref:
            continue
        if ref.startswith("http"):
            resolved.append(ref)
            continue
        normalized = posixpath.normpath(posixpath.join("docs/data/dataset_zoo", ref))
        resolved.append(f"{RAW_BASE}{normalized}")
    return resolved


def main() -> None:
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    IMAGES_DIR.mkdir(parents=True, exist_ok=True)

    markdown = download_text(REPO_URL)
    sections = parse_dataset_sections(markdown)

    catalog: list[dict[str, Any]] = []
    for heading, section_content in sections:
        details = extract_details(section_content)
        description = extract_description(section_content)
        image_urls = extract_images(section_content)
        image_files: list[str] = []
        for image_url in image_urls:
            filename = Path(image_url.split("/")[-1]).name
            destination = IMAGES_DIR / filename
            try:
                download_binary(image_url, destination)
                image_files.append(destination.name)
            except Exception as exc:  # pragma: no cover - network/network errors can happen
                print(f"Skipping image {image_url}: {exc}")

        dataset_name = details.get("dataset_name") or heading
        tags = [tag.strip() for tag in details.get("tags", "").split(",") if tag.strip()]
        catalog.append({
            "name": dataset_name,
            "slug": slugify(dataset_name),
            "heading": heading,
            "description": description,
            "source": details.get("source", ""),
            "size": details.get("size", ""),
            "tags": tags,
            "splits": details.get("splits", ""),
            "images": image_files,
        })

    CATALOG_PATH.write_text(json.dumps(catalog, indent=2), encoding="utf-8")

    summary_lines = [
        "# Voxel51 Dataset Zoo Catalog",
        "",
        f"Captured {len(catalog)} datasets from {REPO_URL}",
        "",
        "## Dataset summary",
        "",
    ]
    for entry in catalog:
        summary_lines.append(f"- {entry['name']} ({', '.join(entry['tags']) or 'untagged'})")
    SUMMARY_PATH.write_text("\n".join(summary_lines) + "\n", encoding="utf-8")

    print(f"Wrote {CATALOG_PATH} and {SUMMARY_PATH}")
    print(f"Downloaded {sum(len(entry['images']) for entry in catalog)} images to {IMAGES_DIR}")


if __name__ == "__main__":
    main()
